Rename non-unique local nonterminals

jfehrle opened the pull request:
This will help us update the grammar in the documentation.

Same description as https://github.com/coq/coq/pull/11009:

The doc_grammar tool doesn't handle multiple local nonterminals with the same name (in different files). productionlists in Sphinx don't handle this either: defining the same nonterminal twice is a fatal error. It will also (in general, though not in this specific case) tend to confuse readers of the documentation. Since the number of such nonterminals is small, I propose we make the names unique rather than enhance the tool.

An added benefit is that it's easier to find the nonterminal when it has a unique name--and less confusing for developers.

Zimmi48 commented on user-contrib/Ltac2/g_ltac2.mlg:
```diff
@@ -643,7 +643,7 @@ GRAMMAR EXTEND Gram
     [ [ c = conversion -> { c } ] ]
   ;
-  orient:
+  orient2:
```
Could be `ltac2_orient`

Zimmi48 commented on vernac/g_vernac.mlg:
```diff
@@ -1137,11 +1137,11 @@ GRAMMAR EXTEND Gram
     [ [ "-" -> { false } | -> { true } ] ]
   ;
-  scope:
+  scope2:
```
What about `search_scope`?

jfehrle replied:
Good idea.

herbelin commented:
That'll be a little more work, but some sharing of the entries could also be tried. At least for `scope`, it is twice the same (and it could actually be called `delimiter`), so it should really be shared. Then for `orient`, sharing would require to add the loc in `g_tactic.mlg`, so it is maybe a bit more disputable. `ltac2_orient` is nice after all.

jfehrle commented:
Updated.  `scope` is now shared.  I renamed it `search_scope` as suggested.  `delimiter` seems too generic.

I think it's better to leave the 2 `orient`s separate in the grammar for somewhat better clarity (but merged in the documentation).  One of them returns true for both "->" and empty, while the other gives distinct values Some true and None for these.  The former would need a match statement to maintain the current behavior.  Is that agreeable?

herbelin commented on vernac/g_vernac.mlg:
```diff
@@ -917,7 +918,7 @@ GRAMMAR EXTEND Gram
 
 GRAMMAR EXTEND Gram
-  GLOBAL: command query_command class_rawexpr gallina_ext;
+  GLOBAL: command query_command class_rawexpr gallina_ext search_scope;
```
Hi, you don't need to add `search_scope` here. You would need to add it to export it only if it were defined in the section.

Otherwise, I agree that `delimiter` is too generic. What about `scope_delimiter` then? If I'm proposing a change, it is because `search_scope` is indeed used as a search scope in `searchabout_query` but not really in `argument_spec`. So, it seems to me that `scope_delimiter` is more neutral and could be a good compromise.

jfehrle commented:
I applied your suggestions.  Perhaps ready to merge?

BTW, I'm looking at later using the `prodn` construct to show grammar in the doc (see https://github.com/coq/coq/issues/11072).  If we do that, the body of `scope_delimiter` would appear inline where it's used and the nonterminal name "scope_delimiter" would not appear in the doc.  In that case, note that picking the best name is only for the benefit of developers.

herbelin commented:
> If we do that, the body of scope_delimiter would appear inline where it's used and the nonterminal name "scope_delimiter" would not appear in the doc. In that case, note that picking the best name is only for the benefit of developers.

Yes, good names are useful for developers too.

@Zimmi48: do you want to assign?

Zimmi48 commented:
> @Zimmi48: do you want to assign?

I could, but this PR is really more one for a member of @coq/parsing-maintainers. You for instance.

herbelin commented on parsing/g_constr.mlg:
```diff
@@ -244,11 +244,11 @@ GRAMMAR EXTEND Gram
-  record_fields:
-    [ [ f = record_field_declaration; ";"; fs = record_fields -> { f :: fs }
+  term_record_fields:
+    [ [ f = record_field_declaration; ";"; fs = term_record_fields -> { f :: fs }
```
@jfehrle: I think we can do better at unifying the syntax of records in `g_constr.mlg` and of records in `g_vernac.mlg`. I propose to do it as part of #11103 as follows:

`G_constr`:
- `record_fields` -> `record_fields_instance`
- `record_field` -> `record_field_instance`

`G_vernac`:
- `record_fields` kept unchanged

The motivation is that `record_fields` in `G_constr` is actually the syntax for the inhabitant of a record type, while in `G_vernac` it is about the record type itself.

jfehrle replied:
Done, though `record_field` -> `record_field_instance` is in g_vernac.

herbelin commented on plugins/ltac/g_tactic.mlg:
```diff
@@ -450,11 +450,11 @@ GRAMMAR EXTEND Gram
         ":"; ty=lconstr; ")" -> { (loc, id, bl, ann, ty) } ] ]
   ;
-  fixannot:
+  ecase_fixannot:
```
I don't think `ecase` makes sense here. I don't see where `fixdecl` is used elsewhere. So, I propose to keep `fixdecl` (or at worst to make it `fixtac_decl`).

As for `ecase_fixannot`, I would turn it e.g. into `structannot` or `fixtac_annot`.

jfehrle replied:
Maybe commenting on the wrong line?  `fixannot` is defined in both `g_constr` and `g_tactic`.  I'll change `ecase_fixdecl` back to `fixdecl`.

jfehrle replied:
and changed `ecase_fixannot`

herbelin commented on vernac/g_vernac.mlg:
```diff
@@ -490,8 +491,8 @@ GRAMMAR EXTEND Gram
-  record_fields:
-    [ [ f = record_field; ";"; fs = record_fields -> { f :: fs }
+  vernac_record_fields:
+    [ [ f = record_field; ";"; fs = vernac_record_fields -> { f :: fs }
```
See above.

jfehrle commented:
Updated.

herbelin commented:
The new renaming is very good, except the renaming of `record_fields` in `g_vernac.mlg` which I don't understand. The renaming of `record_fields` done in `g_constr.mlg` should be enough to avoid the clash, right? So, the one in `g_vernac.mlg` should be cancelled (technically, since `record_fields` in `g_vernac.mlg` is actually about the record type and `record_fields` in `g_constr.mlg` is about the values of a record type, we really should not call the one in `g_vernac.mlg` an "instance").

jfehrle commented:
Your suggestion was a bit unclear, I was guessing what you really wanted.

> the renaming of `record_fields` in `g_vernac.mlg`

That was not renamed.  `record_field` (no "s") was renamed to `record_field_instance`.  You want me to reverse that one.  Right?

Updated.

herbelin commented:
Indeed, I was not clear at all. In my yesterday comment, I meant `G_constr.record_field_declaration`-> `G_constr.record_field_instance` (in addition to `G_constr.record_fields_declaration`-> `G_constr.record_fields_instance`) and in my next message, I meant the renaming of `record_field` (w/o `s`). Sorry about my confusion.

jfehrle commented:
OK, I think I got it.  Please check.  Also this version compiles :-)

herbelin commented:
Wonderful. I shall merge tomorrow morning (Paris time).
